### PR TITLE
Initial JA3 integration

### PR DESCRIPTION
**Pull request by thus**
This PR contains an initial JA3 integration (https://github.com/salesforce/ja3). This means that there are some features that are not done yet, and that will (hopefully) be added in the future, but it contains the core functionality.

Features implemented:
* Decode more of the TLS CLIENT_HELLO packet, and generate a JA3 string and a JA3 hash (md5sum of the JA3 string).
* Detection (mpm) keyword 'ja3_hash' for matching on the JA3 hash.
* Detection (mpm) keyword 'ja3_str' for matching on the entire or parts of the JA3 string.
* JSON logger for JA3.
* JA3 metadata in alerts.
* Lua functions 'Ja3GetHash' and 'Ja3GetStr' to access the JA3 buffers from Lua scripts.
* 'ja3_hash' and 'ja3_str' as custom fields in the TLS JSON logger.

As I briefly discussed with @inliniac, I also want to implement 'ja3_desc'. To do this I want to read a JSON file with JA3 fingerprints (hash, string and description) when Suricata starts and on rule reload. This file should contain all the known fingerprints and enable us to match on part of the description, e.g. "ja3_desc; content:"Firefox";). It should also enable us to match on all unknown JA3 fingerprints. Since the de_ctx is not available when decoding the TLS packet, I have not added this feature yet, but it is something that (IMO) would add a lot of value to the JA3 integration.

As of now, I have choose to disable generating JA3 fingerprints by default. I have added an entry to enable it to the suricata.yaml file in the repo, but this means that it won't be enabled before the user updates their Suricata config. Please let me know if you disagree on doing this.

The authors of JA3 is presenting on Shmoocon in late January, so it would be really cool if we could have JA3 support in Suricata by then ;)

Redmine:
https://redmine.openinfosecfoundation.org/issues/2192

Prscript:
- PR thus-pcap: https://buildbot.openinfosecfoundation.org/builders/thus-pcap/builds/137
- PR thus: https://buildbot.openinfosecfoundation.org/builders/thus/builds/137

### CODE REVIEW

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -258,55 +263,214 @@ static int SSLSetTxMpmIDs(void *vtx, uint64_t mpm_ids)
...
+
+        JA3Buffer *ja3_cipher_suites = Ja3BufferInit();
+        if (ja3_cipher_suites == NULL)
+            return -1;
+
+        uint16_t processed_len = 0;
+        while (processed_len < cipher_suites_length)
+        {
+            if (!(HAS_SPACE(2)))
+                goto invalid_length;
```
Comment (victorjulien) [inline src/app-layer-ssl.c]: ja3_cipher_suites memleak

File: src/app-layer-ssl.c [hunk 2]
```diff
@@ -258,55 +263,214 @@ static int SSLSetTxMpmIDs(void *vtx, uint64_t mpm_ids)
...
+        SSLSetEvent(ssl_state,
+                    TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+        return -1;
+    }
+
+    /* Skip compression methods */
+    uint8_t compression_methods_length = **input;
+    *input += 1;
+
+    *input += compression_methods_length;
```
Comment (victorjulien) [inline src/app-layer-ssl.c]: should we first have a check for HAS_SPACE(compression_methods_length)?

File: src/app-layer-ssl.c [hunk 3]
```diff
@@ -378,30 +543,186 @@ static int TLSDecodeHandshakeHello(SSLState *ssl_state, uint8_t *input,
...
-                memcpy(ssl_state->curr_connp->sni, input, sni_strlen - 1);
+                memcpy(ssl_state->curr_connp->sni, *input, sni_strlen - 1);
                 ssl_state->curr_connp->sni[sni_strlen-1] = 0;
 
-                input += sni_len;
+                *input += sni_len;
+                break;
+            }
+
+            case SSL_EXTENSION_ELLIPTIC_CURVES:
```
Comment (victorjulien) [inline src/app-layer-ssl.c]: I thnk it would make sense to break up this function/loop as it's quite long

File: src/util-ja3.c [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+/**
+ * \internal
+ * \brief Resize buffer if it is full.
+ *
+ * \param buffer The buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+static int Ja3BufferResizeIfFull(JA3Buffer *buffer, unsigned len)
```
Comment (victorjulien) [inline src/util-ja3.c]: please use explicit type instead of just 'unsigned', so something like uint32_t

**Comment by victorjulien** [comment]
Can you post examples of the tls eve addition and the ja3 eve records to this PR?

File: src/app-layer-ssl.c (hunk 2 above)
Comment (thus) [reply src/app-layer-ssl.c]: Do we need it? We are not dereferencing input as far as I can see.

File: src/app-layer-ssl.c (hunk 1 above)
Comment (thus) [reply src/app-layer-ssl.c]: Will fix.

File: src/app-layer-ssl.c (hunk 3 above)
Comment (thus) [reply src/app-layer-ssl.c]: I'll create some new functions for you :)

File: src/util-ja3.c (hunk 4 above)
Comment (thus) [reply src/util-ja3.c]: Will do.

**Comment by thus** [comment]
Example of JA3 fields added to TLS log:

```json
{
  "timestamp": "2017-11-28T20:52:15.413711+0100",
  "flow_id": 248041851312109,
  "pcap_cnt": 10,
  "event_type": "tls",
  "src_ip": "192.168.86.32",
  "src_port": 60596,
  "dest_ip": "192.168.31.104",
  "dest_port": 443,
  "proto": "TCP",
  "tls": {
    "session_resumed": true,
    "sni": "something.com",
    "ja3_hash": "2efb07037a97b06201ab4fe7ec0c326e",
    "ja3_str": "771,49195-49199-52393-52392-49196-49200-49162-49161-49171-49172-51-57-47-53-10,0-23-65281-10-11-35-16-5-13-21,29-23-24-25,0"
  }
}
```

Example of the same entry, but in JA3 log:

```json
{
  "timestamp": "2017-11-28T20:52:15.413711+0100",
  "flow_id": 248041851312109,
  "pcap_cnt": 10,
  "event_type": "ja3",
  "src_ip": "192.168.86.32",
  "src_port": 60596,
  "dest_ip": "192.168.31.104",
  "dest_port": 443,
  "proto": "TCP",
  "ja3": {
    "hash": "2efb07037a97b06201ab4fe7ec0c326e",
    "str": "771,49195-49199-52393-52392-49196-49200-49162-49161-49171-49172-51-57-47-53-10,0-23-65281-10-11-35-16-5-13-21,29-23-24-25,0"
  }
}
```

The last example (the 'ja3' tag) is how it shows up as meta data to an alert as well.

File: src/app-layer-ssl.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-ssl.c]: it's moved forward for the next caller isn't it? Seems unsafe to do this w/o check?

Comment (thus) [reply src/app-layer-ssl.c]: Yes, it is. I'll add a check if that makes you sleep better at night ;)

**Comment by thus** [comment]
Replaced by #3141

**Pull request closed without merging**